File: Gestion de la Informacion/Practicas finalizadas/Practica4/src/Usuario.java
import java.util.*;

public class Usuario implements Comparable<Usuario>
{
    private static String BD_SERVER = "localhost";
    private static String BD_NAME = "GI1920";
    
    private String nif;
    private String password;
    private Rol rol;

	public static List<Usuario> ListaUsuarios()
	{
		// Retorna una lista con todos los obejtos de la clase almacenados en la base de datos
		BD miBD = new BD(BD_SERVER,BD_NAME);
		ArrayList<Usuario> lista = new ArrayList<Usuario>(); 
					
		for(Object[] tupla: miBD.Select("SELECT nif,password FROM tUsuario;"))
		{
			String nif = (String)tupla[0];
			String p = (String)tupla[1];
			Usuario u = new Usuario(nif,p);
			lista.add(u);
		}
		return lista;
	}
	
    public Usuario(String n, String p)
    {
		// Crea el objeto cargando sus valores de la base de datos
    	try
    	{
	    	BD miBD = new BD(BD_SERVER,BD_NAME);			
	  
			Object[] tupla = miBD.Select("SELECT * FROM tUsuario " 
					+ "WHERE nif='"+n+ "';").get(0);
			
	        nif = (String)tupla[0];
	        password = (String)tupla[1];
	        
	        if (!password.equals(p))
	        {
	        	nif = password = null;
	        	throw new BDException("Error: Usuario o Contraseña incorrecto");
	        }
	        rol = new Rol((String)tupla[2]);        
    	}
    	catch (Exception ex)
    	{
    		throw new BDException("Error: Usuario o Contraseña incorrecto");
    	}
    }
    
    public Usuario(String n, String p, Rol r)
    {
		// Crea el objeto y lo inserta en la base de datos
    	BD miBD = new BD(BD_SERVER,BD_NAME);		
    	
    	miBD.Insert("INSERT INTO tUsuario VALUES("
				+ "'" + n + "', '" + p + "', '" + r.getRolName() + "');" );			
	
        nif = n;
        password = p;
        rol = r;
    }

    public String getNIF() 
    { 
    	return nif; 
    }

    public void setNIF(String value) 
    { 
    	
		// Actualiza el atributo en memoria y en la base de datos
    	BD miBD = new BD(BD_SERVER,BD_NAME);
    	miBD.Update("UPDATE tUsuario SET nif = '" + value 
    			+ "' WHERE nif ='"+ this.nif + "';");
    	
    	nif = value; 
    }

    public void borraUsuario() 
    {     	
		// Actualiza el atributo en memoria y en la base de datos
    	BD miBD = new BD(BD_SERVER,BD_NAME);
    	miBD.Delete("DELETE FROM  tUsuario WHERE nif ='"+ this.nif + "';");    	
    	nif = password = null;
    	rol = null;
    }
    public String getPassword() 
    { 
    	return password; 
    }
        
    public void setPassword (String value)
    { 
		// Actualiza el atributo en memoria y en la base de datos
    	BD miBD = new BD(BD_SERVER,BD_NAME);
    	miBD.Update("UPDATE tUsuario SET password = '" + value 
    			+ "' WHERE nif ='"+ this.nif + "';");
    	
    	password = value; 
    }

    public Rol getRol()
    {
        return rol;
    }
    
    public void setRol()
    {
    	throw new BDException("Error: Un usuario no puede cambiar el rol de si mismo.");     
    }

    public void ModiRol(Usuario u, Rol r)
    {
        if (this.rol.getAdmin()) 
        {
		// Actualiza el atributo rol de u en memoria y en la base de datos
        	BD miBD = new BD(BD_SERVER,BD_NAME);
        	miBD.Update("UPDATE tUsuario SET rolName = '" + r.getRolName() 
        			+ "' WHERE nif ='"+ u.getNIF() + "';");
        	
        	 u.rol = r;
        	 
        }
        else throw new Error("El usuario " + this.getNIF() 
        		+ " no puede cambiar el rol del usuario " 
        		+ u.getNIF());
    }


    public boolean AccesoPantalla(String p)
    {
        return rol.Acceso(p);
    }

    public boolean InsertarPantalla(String p)
    {
        return rol.Insertar(p);
    }

    public boolean ModificarPantalla(String p)
    {
        return rol.Modificar(p);
    }

    public boolean BorrarPantalla(String p)
    {
        return rol.Borrar(p);
    }

    public String toString()
    {
        return nif + ";" + password + ";" + rol;
    }

    public boolean equals(Object obj)
    {
        return (obj instanceof Usuario) && ((Usuario)obj).nif.equals(this.nif);
    }

    public int hashCode()
    {
        return this.nif.hashCode();
    }

	public int compareTo(Usuario u) 
	{
		return this.nif.compareToIgnoreCase(u.nif);	
	}
}

File: Gestion de la Informacion/alumnosP6/Usuario.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using BDLibrary;

namespace farmacia
{
    public class Usuario
    {
        private static string BD_SERVER = Properties.Settings.Default.BD_SERVER;
        private static string BD_NAME = Properties.Settings.Default.BD_NAME;


        private string nif;
        private string password;
        private Rol rol;


	    public static List<Usuario> ListaUsuarios()
	    {
		    List<Usuario> lista = new List<Usuario>();

            SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);

            foreach (object[] tupla in miBD.Select("SELECT nif, password FROM tUsuario;"))
            {
                string n = (string)tupla[0];
                string p = (string)tupla[1];
                lista.Add(new Usuario(n,p));
            }

		    return lista;
	    }

        public Usuario(string n, string p)
        {

            try
            {
                SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
                object[] tupla = miBD.Select("SELECT * FROM tUsuario WHERE nif = '" + n + "';")[0];

                nif = (string)tupla[0];
                password = (string)tupla[1];

                if (!password.Equals(p))
                {
                    nif = password = null;
                    throw new Error("Usuario o Contraseña Incorrecta: ");
                }

                rol = new Rol((string)tupla[2]);
            }
            catch
            {
                throw new Error("Usuario o Contraseña Incorrecta: ");
            }
                
        }

        public Usuario(string n, string p, Rol r)
        {

            SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);

            miBD.Insert("INSERT INTO tUsuario VALUES('" + n + "', '"
                + p + "', '" + r.RolName + "');");

            nif = n;
            password = p;
            rol = r;

        }

        public string NIF
        {
            get { return nif; }
            set {
                    SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
                    miBD.Update("UPDATE tUsuario SET nif = '" + value
                            + "' WHERE nif = '" + nif + "';");
                    nif = value;
                }
        }

        public string Password
        {
            get { return password; }
            set {
                    SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
                    miBD.Update("UPDATE tUsuario SET password = '" + value
                            + "' WHERE nif = '" + nif + "';");
                    password = value;
                }
        }

        public Rol MyRol
        {
            get { return rol; }
        }

        public void ModiRol(Usuario u, Rol r)
        {
            if (!this.rol.Admin)
                throw new Error("El usuario " + this.NIF
                    + " no puede cambiar el rol del usuario " + u.NIF);
            SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
            miBD.Update("UPDATE tUsuario SET rolName = '" + r.RolName
                    + "' WHERE nif = '" + u.NIF + "';");
            u.rol = r;

        }

        public void BorrarUsuario()
        {
            SQLSERVERDB miBD = new SQLSERVERDB(BD_SERVER, BD_NAME);
            miBD.Delete("DELETE tUsuario WHERE nif = '" + nif + "';");
            nif = password = null;
            rol = null;
        }


        public bool AccesoPantalla(string p)
        {
            return rol.Acceso(p);
        }

        public bool InsertarPantalla(string p)
        {
            return rol.Insertar(p);
        }

        public bool ModificarPantalla(string p)
        {
            return rol.Modificar(p);
        }

        public bool BorrarPantalla(string p)
        {
            return rol.Borrar(p);
        }

        public override string ToString()
        {
            return nif + ";" + password + ";" + rol;
        }

        public override bool Equals(object obj)
        {
            return obj is Usuario && ((Usuario)obj).NIF.Equals(NIF);
        }

        public override int GetHashCode()
        {
            return NIF.GetHashCode();
        }
    }
}
